LocalPackages/DataBrokerProtection/Sources/DataBrokerProtection/Model/DBPUICommunicationModel.swift

```diff
@@ -146,11 +146,93 @@ struct DBPUIDataBrokerList: DBPUISendableMessage {
     let dataBrokers: [DBPUIDataBroker]
 }
 
-/// Message Object representing a requested change to the user profile's brith year
+/// Message Object representing a requested change to the user profile's birth year
 struct DBPUIBirthYear: Codable {
     let year: Int
 }
 
+/// Message Object representing a supported timeline event
+/// https://app.asana.com/0/481882893211075/1208663928051302/f
+struct DBPUITimelineEvent: Codable {
+    enum EventType: String, Codable {
+        case recordFound = "record-found"
+        case recordReappeared = "record-reappeared"
+        case optOutSubmitted = "opt-out-submitted"
+        case estimatedRemoval = "estimated-removal"
+        case removed = "removed"
+    }
+
+    let type: EventType
+    let date: Double
+
+    var eventDate: Date {
+        Date(timeIntervalSince1970: date)
+    }
+
+    init?(type: EventType, date: Date?) {
+        guard let date else { return nil }
+        self.type = type
+        self.date = date.timeIntervalSince1970
+    }
+}
+
+extension DBPUITimelineEvent {
+    init?(foundDate: Date?) {
+        self.init(type: .recordFound, date: foundDate)
+    }
+
+    init?(reappearedDate: Date?) {
+        self.init(type: .recordReappeared, date: reappearedDate)
+    }
+
+    init?(optOutSubmittedDate: Date?) {
+        self.init(type: .optOutSubmitted, date: optOutSubmittedDate)
+    }
+
+    init?(estimatedRemovalDate: Date?) {
+        self.init(type: .estimatedRemoval, date: estimatedRemovalDate)
+    }
+
+    init?(removedDate: Date?) {
+        self.init(type: .removed, date: removedDate)
+    }
+
+    static func from(historyEvents: [HistoryEvent], removedDate: Date?) -> [DBPUITimelineEvent] {
+        var timelineEvents = historyEvents.compactMap { event in
+            switch event.type {
+            case .matchesFound:
+                return DBPUITimelineEvent(foundDate: event.date)
+            case .reAppearence:
+                return DBPUITimelineEvent(reappearedDate: event.date)
+            case .optOutRequested:
+                return DBPUITimelineEvent(optOutSubmittedDate: event.date)
+            default:
+                return nil
+            }
+        }
+
+        let mostRecentFoundEvent = timelineEvents.filter({ $0.type == .recordFound || $0.type == .recordReappeared }).max(by: <)
+        let mostRecentOptOutSubmittedEvent = timelineEvents.filter({ $0.type == .optOutSubmitted }).max(by: <)
+
+        if let optOutSubmittedDate = (mostRecentOptOutSubmittedEvent ?? mostRecentFoundEvent)?.eventDate,
+           let estimatedRemovalEvent = DBPUITimelineEvent(estimatedRemovalDate: Calendar.current.date(byAdding: .day, value: 14, to: optOutSubmittedDate)) {
+            timelineEvents.append(estimatedRemovalEvent)
+        }
+
+        if let removedEvent = DBPUITimelineEvent(removedDate: removedDate) {
+            timelineEvents.append(removedEvent)
```

Comment:
Am I right in thinking that if the entry has been removed, then the timeline will have an estimated removal date event, followed by an actual removed event? Is that what we want?

Reply:
Ah good point. Logically we shouldn't have the estimated removal date when the profile has been removed 😅